examples/vision/depth_estimation.py

```diff
@@ -0,0 +1,516 @@
+        self.reluA = layers.LeakyReLU(alpha=0.2)
+        self.reluB = layers.LeakyReLU(alpha=0.2)
+        self.bn2a = tf.keras.layers.BatchNormalization()
+        self.bn2b = tf.keras.layers.BatchNormalization()
+
+        self.pool = layers.MaxPool2D((2, 2), (2, 2))
+
+    def call(self, input_tensor):
+        d = self.convA(input_tensor)
+        x = self.bn2a(d)
+        x = self.reluA(x)
+
+        x = self.convB(x)
+        x = self.bn2b(x)
+        x = self.reluB(x)
+
+        x += d
+        p = self.pool(x)
+        return x, p
+
+
+class UpscaleBlock(layers.Layer):
+    def __init__(self, filters, kernel_size=(3, 3), padding="same", strides=1):
+        super().__init__()
+        self.us = layers.UpSampling2D((2, 2))
+        self.convA = layers.Conv2D(filters, kernel_size, strides, padding)
+        self.convB = layers.Conv2D(filters, kernel_size, strides, padding)
+        self.reluA = layers.LeakyReLU(alpha=0.2)
+        self.reluB = layers.LeakyReLU(alpha=0.2)
+        self.bn2a = tf.keras.layers.BatchNormalization()
+        self.bn2b = tf.keras.layers.BatchNormalization()
+        self.conc = layers.Concatenate()
+
+    def call(self, x, skip):
+        x = self.us(x)
+        concat = self.conc([x, skip])
+        x = self.convA(concat)
+        x = self.bn2a(x)
+        x = self.reluA(x)
+
+        x = self.convB(x)
+        x = self.bn2b(x)
+        x = self.reluB(x)
+
+        return x
+
+
+class BottleNeckBlock(layers.Layer):
+    def __init__(self, filters, kernel_size=(3, 3), padding="same", strides=1):
+        super().__init__()
+        self.convA = layers.Conv2D(filters, kernel_size, strides, padding)
+        self.convB = layers.Conv2D(filters, kernel_size, strides, padding)
+        self.reluA = layers.LeakyReLU(alpha=0.2)
+        self.reluB = layers.LeakyReLU(alpha=0.2)
+
+    def call(self, x):
+        x = self.convA(x)
+        x = self.reluA(x)
+        x = self.convB(x)
+        x = self.reluB(x)
+        return x
+
+
+def UnetModel(height, width):
+
+    f = [16, 32, 64, 128, 256]
+
+    inputs = layers.Input((height, width, 3))
+
+    c1, p1 = DownscaleBlock(f[0])(inputs)
+    c2, p2 = DownscaleBlock(f[1])(p1)
+    c3, p3 = DownscaleBlock(f[2])(p2)
+    c4, p4 = DownscaleBlock(f[3])(p3)
+
+    bn = BottleNeckBlock(f[4])(p4)
+
+    u1 = UpscaleBlock(f[3])(bn, c4)
+    u2 = UpscaleBlock(f[2])(u1, c3)
+    u3 = UpscaleBlock(f[1])(u2, c2)
+    u4 = UpscaleBlock(f[0])(u3, c1)
+
+    outputs = layers.Conv2D(1, (1, 1), padding="same", activation="tanh")(u4)
+
+    model = tf.keras.Model(inputs, outputs)
+    return model
+
+
+"""
+## Optimizing Loss
+We have tried to optimize 3 losses in our model.
+1. Structural similarity index(SSIM).
+2. L1-loss, or Point-wise depth in our case.
+3. Edge wide depth with depth smoothness.
+
+Out of the three loss functions SSIM contributed the most in improving model performance.
+"""
+
+
+class DepthEstimationModel(tf.keras.Model):
```

Comment:
In my recent commit/ updated code I have updated everything as you suggested except for this part. I found it a bit difficult to combine them into a single subclass model in my case, which maybe because of my lack of knowledge on that. if this creates an issue for approval could you please suggest to me an example so that I could understand it in a better way? i followed Keras examples like [image captioning](https://keras.io/examples/vision/image_captioning/) and [oxford_pets_image_segmentation](https://keras.io/examples/vision/oxford_pets_image_segmentation/)